### DISCUSSION TITLE
Lazy loading and validating value

### DISCUSSION THREAD

**Discussion by mtvx** [pydantic/pydantic#9846] [General]
Hey, we have a traditional dict-based config, which may contain magic values:

```json
{
  "field1": "usual field",
  "field2": {
    "magic:value": {
      "secret_arn": "arn:..."
    }
  }
}
```

When that `{"magic:value": ...}` value is faced, the actual data is looked up from an AWS Secret (or it could be DB, or anything).

The resolution is now lazy, and I'd like to retain it that way.

In order to map the dict into Pydantic model, I'd sketch something like:

```py
class Settings(BaseSettings):
    field1: str
    field2: str
```

What should happen is that:

- Feeding the above dict should not raise validation error. I think I could handle this with a `WrapValidator`
- Accessing the field (`field2`) for the first time, should resolve and cache the value, validating against the correct type (`str` and any extra validations that the field might have). I think I could handle this with `@property` accessor (and naming the field differently, but then use `alias`).

But all that adds quite a lot of boilerplate, and it's somewhat confusing. Plus it would need to be repeated for each field. What would be a good _generic_ solution, to support the lazy evaluation/validation anywhere in the dict? A simple `__getitem__` won't do, because I also need to skip the validation at the start, and then I want 100% perfect typing for the accessor (ie. all fields need to be declared in the class).

Note that I don't want to do:

```py
class MagicValue(BaseSettings):
    secret_arn: str

class Settings(BaseSettings):
    field1: str
    field2: str | MagicValue
```

Because then accessing the settings `.field2` would be union of `str | MagicValue` instead of just `str`.

Any pointers?

**Comment by gustavosett** [comment]
Try creating a custom descriptor for Lazy Loading

```py
class LazyField:
    def __init__(self, name=None, type_=str):
        self.name = name
        self.type_ = type_
        self.value = None
        self.is_loaded = False

    def __get__(self, obj, objtype=None):
        if obj is None:
            return self
        if not self.is_loaded:
            self.value = self.load_value(obj._config[self.name])
            self.is_loaded = True
        return self.value

    def __set__(self, obj, value):
        raise AttributeError("Cannot set value directly, use the config dictionary.")

    def load_value(self, value):
        # implement lazy loading logic here, possibly checking for magic values and fetching secrets
        if isinstance(value, dict) and 'magic:value' in value:
            # as example a function `resolve_secret` fetches and returns the secret value
            resolved_value = resolve_secret(value['magic:value']['secret_arn'])
        else:
            resolved_value = value

        # validate the resolved value if necessary
        # this example assumes all values should be strings after resolution
        if not isinstance(resolved_value, self.type_):
            raise ValueError(f"Expected value of type {self.type_}, got {type(resolved_value)}")
        return resolved_value
```

You can use like that:

```py
from pydantic import BaseModel, BaseSettings

class Settings(BaseSettings):
    class Config:
        arbitrary_types_allowed = True

    def __init__(self, **data):
        super().__init__(**data)
        self._config = data
        # initialize each LazyField descriptor with the field name
        for field_name, field_type in self.__annotations__.items():
            descriptor = LazyField(name=field_name, type_=field_type)
            setattr(self.__class__, field_name, descriptor)

# example usage
config_data = {
    "field1": "usual field",
    "field2": {
        "magic:value": {
            "secret_arn": "arn:aws:secretsmanager:..."
        }
    }
}

settings = Settings(**config_data)
print(settings.field1)  # Output: usual field
print(settings.field2)  # Output: resolved and validated secret value
```

**Reply by mtvx** [reply]
Hey, sorry somehow I've totally missed your response. How would you actually define `field1` and `field2`? The code would just raise in `super().__init__(**data)` if you pass dict in place of `str` annotated field, no?

Also, it seems to modify the _class_ fields, not instance fields.

**Comment by EytanDn** [comment]
Hey! Did you manage this? looking into implementing solution in my codebase!

**Reply by mtvx** [reply]
I did manage to implement it, but it's a long and ugly kludge, which I would rather not use.

My suggestion would be to NOT to do it, and choose a simpler model, e.g. eagerly resolve the values (so your application will also fail on startup and not runtime, should resolving some value fail) 🤔
